spec/sidekiq_flow/adapters/set_storage_spec.rb

```diff
@@ -0,0 +1,480 @@
+
+      before do
+        described_class.start_workflow(workflow)
+      end
+
+      it 'should NOT create a new redis key' do
+        expect {
+          subject
+        }.not_to change { redis.keys.count }
+      end
+
+      it 'should ONLY have 2 redis key' do
+        subject
+        expect(redis.keys.count).to eq(2)
+      end
+
+      it 'should override the existing stored workflow' do
+        hash1 = redis.hgetall(redis.keys.last)
+        workflow.clear_branch!('TestTask1')
+        subject
+        hash2 = redis.hgetall(redis.keys.last)
+
+        expect(hash1).not_to eq(hash2)
+
+      end
+    end
+  end
+
+  describe '.store_task' do
+    context 'success' do
+      subject { described_class.store_task(task) }
+      let(:task) { workflow.tasks.first }
+
+      before do
+        described_class.start_workflow(workflow)
+      end
+
+      it 'should store task info in redis' do
+        expect {
+          task.fail!
+          subject
+        }.to change {
+          SidekiqFlow::Client.find_task(workflow.id, task.class.name).status
+        }.from('enqueued').to('failed')
+      end
+
+      context 'NOT yet succeeded workflow' do
+        it 'should move workflow key to finished set' do
+          subject
+          expect( redis.sismember('in-progress', SidekiqFlow::Client.find_workflow_key(workflow.id)) ).to eq(true)
+        end
+      end
+    end
+  end
+
+  describe '.find_workflow' do
+    subject { described_class.find_workflow(workflow.id) }
+
+    context 'success' do
+      before do
+        described_class.start_workflow(workflow)
+      end
+
+      it 'should return a Workflow instance' do
+        result = subject
+        expect(result).to be_a_kind_of(TestWorkflow)
+      end
+
+      it 'should return a the correct Workflow instance' do
+        result = subject
+        expect(result.id).to eq(workflow.id)
+      end
+    end
+
+    context 'exception' do
+      it 'should raise WorkflowNotFound' do
+        expect {
+          subject
+        }.to raise_exception(SidekiqFlow::WorkflowNotFound)
+      end
+    end
+  end
+
+  describe '.destroy_workflow' do
+    subject { described_class.destroy_workflow(workflow.id) }
+
+    context 'success' do
+      before do
+        described_class.start_workflow(workflow)
+      end
+
+      it 'should delete the workflow' do
+        expect {
+          subject
+        }.to change { redis.keys.count }.from(2).to(1)
+
+        expect( redis.sismember('finished', SidekiqFlow::Client.find_workflow_key(workflow.id)) ).to eq(false)
+      end
+    end
+  end
+
+  describe '.destroy_succeeded_workflows' do
+    let(:workflow2) { TestWorkflow.new(id: 345) }
+
+    subject { described_class.destroy_succeeded_workflows }
+
+    context 'success' do
+      before do
+        described_class.start_workflow(workflow)
+        SidekiqFlow::Worker.drain
+      end
+
+      it 'should ONLY delete succeeded workflows' do
+        described_class.start_workflow(workflow2)
```

Comment:
Move `described_class.start_workflow(workflow2)` to the before block.

There are too make checks, but is very hard to follow, let's test one thing at a time (extract to multiple specs).

e.g 

```ruby
it 'changes redis key count' do
  expect {
    subject
  }.to change {
    redis.keys.count)
  }.from(4).to(2)
end
```